blob/gcsblob/gcsblob.go

```diff
@@ -230,6 +306,56 @@ func OpenBucket(ctx context.Context, client *gcp.HTTPClient, bucketName string,
 	return blob.NewBucket(drv), nil
 }
 
+// credentialsClient is a wrapper for lazy initialization.
+type credentialsClient struct {
+	init   sync.Once
+	client *credentials.IamCredentialsClient
+	err    error
+}
+
+// errPassthrough satisfies the signature expected by storage.SignedURLOptions.SignBytes
+// but really just passes through any permanent error.
+// It's extracted to minimize allocations.
+func (c *credentialsClient) errPassthrough(_ []byte) ([]byte, error) {
+	return nil, c.err
+}
+
+// CreateMakeSignBytesWith produces a MakeSignBytes variant from an expanded parameter set.
+// It essentially adapts a remote call to the IAM Credentials API
+// to the function signature expected by storage.SignedURLOptions.SignBytes.
+func (c *credentialsClient) CreateMakeSignBytesWith(lifetimeCtx context.Context, googleAccessID string) func(context.Context) SignBytesFunc {
+	return func(requestCtx context.Context) SignBytesFunc {
+		c.init.Do(func() {
+			c.client, c.err = credentials.NewIamCredentialsClient(lifetimeCtx)
```

Comment:
What's this context used for in NewIamCredentialsClient? I am pretty sure it can be cancelled already by this point since it's being called lazily... Maybe better to use a Background instead?

Reply:
You're right, but I believe the Background context should be passed further up instead and this left as-is: To enable users to change the AccessID used I'm considering making the `credentialsClient` public (capital `CredentialsClient`; consequently we'd not remove the currently redundant `Close`). Users will be able to instantiate their own. Or at least move it to its own file to make extraction easier.

Options I'd move on: (1) keep it private, remove and simplify as currently suggested by you; (2) private, but in its own file with everything in place; (3) make it public. I'm for (3), but it's no module owned by me.

Reply:
I like the idea of making it public so that users can do this on their own if they want to.

I don't think the lazy initialization is adding much value and it's adding significant complexity; so maybe we can just do a func something like:

```
func CredentialsSigner(client *credentials.IamCredentialsClient, googleAccessID string) func(context.Context) SignBytesFunc {
  ...
}
```

You can initialize the client in the default lazyCredsOpener, which already has a init.Once and mechanism for persistently returning errors, and it can use the passed-in context there to instantiate the client. 

Then there's no need for the credentialsClient struct at all.

Users who want to specify a googleAccessID can initialize (and Close) the client on their own.

Reply:
My own use-case is, I have this on AppEngine as one part of many, and I'd like to defer anything expensive like establishing remote connections. Further, a user might override the defaults we are about to set and the connection would've been made needlessly.

I chose to believe you'll like the new implementation, especially as it adds value by being reusable (at the expense of the user managing it, i. e. closing it) and serving as example for how to employ a HSM in an on-prem environment.

As to not further scope-creep of this patch, I'll prepare this now but petition to make it public/export in a separate PR with the method we currently won't use (`Close`).